accord-core/src/main/java/accord/local/CommandStores.java

```diff
@@ -65,19 +67,23 @@ CommandStores<?> create(NodeTimeService time,
         private final DataStore store;
         private final ProgressLog.Factory progressLogFactory;
         private final CommandStore.Factory shardFactory;
+        private final RandomSource random;
 
-        Supplier(NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, CommandStore.Factory shardFactory)
+        Supplier(NodeTimeService time, Agent agent, DataStore store, RandomSource random, ProgressLog.Factory progressLogFactory, CommandStore.Factory shardFactory)
         {
             this.time = time;
             this.agent = agent;
             this.store = store;
+            this.random = random;
             this.progressLogFactory = progressLogFactory;
             this.shardFactory = shardFactory;
         }
 
         CommandStore create(int id, RangesForEpochHolder rangesForEpoch)
         {
-            return shardFactory.create(id, time, agent, store, progressLogFactory, rangesForEpoch);
+            CommandStore store = shardFactory.create(id, time, agent, this.store, progressLogFactory, rangesForEpoch);
+            store.execute(() -> CommandStore.register(store));
```

Comment:
it could be, but then causes duplication of logic.  As executors are expected to be ordered, this makes sure that the first action is to register and we avoid having to handle in each store uniquely

The `Synchronized` has to redo all this logic as the register makes no sense in that case, rest are all thread based so avoids duplication.

Now, if we have `AccordCommandStore` extend `SingleThread` this would lower the amount of duplication as there are only 2 real implementations in this case... I don't remember why those classes were split.

Reply:
My concern isn't code reuse, but that this is _explicitly_ an implementation detail of the `CommandStore` that has now leaked into `CommandStores`, in a manner that is very far from obvious, and certainly brittle. If you want to make `CommandStore` an abstract class to implement this behaviour, or make a threaded `CommandStore` that does this for others to extend, that would be great of course.

Reply:
Ok I hope I have a solution that works here...

I added a new `register` method to `CommandStore` which `CommandStores` calls on init.  `CommandStore.register` defaults to this same logic and sync store overrides as a noop...

Why did I go this route?  I need to run in the `Executor` once and have access to `this`, but if I do that in the constructor I could access a non-finalized `CommandStore`, so felt its best to do this logic outside of the constructor... which means I need something to trigger...

This makes it so `CommandStores` doesn't know "how" we register, it only knows that we "need" to register...  since `CommandStores` no longer knows the "how" then it doesn't know this implementation detail....

Reply:
I'm a bit confused at this, taking a closer look. The `SingleThread` implementation seems to not override register, even though it seemingly has no reason to use the `Unsafe` implementation, since it is able to just consult its `Thread` register, as this is of course much simpler and cheaper.

However, the `register` method only makes logical sense for the `SingleThread` implementations? The others need to surely maintain it on entry/exit?

Reply:
The default is `execute(() -> CommandStore.register(this));` which adds the current to the shared `ThreadLocal`; this default is used by `AccordCommandStore`, `SingleThread`, and `Debug` (as it extends `SingleThread`).

There are 2 non-threaded based `CommandStore`s that mark this method as a no-op: that is `Synchronized` and `DelayedCommandStore`

> The SingleThread implementation seems to not override register, even though it seemingly has no reason to use the Unsafe implementation, since it is able to just consult its Thread register, as this is of course much simpler and cheaper.

What do you mean by `Unsafe`?  The current logic calls `CommandStore.register(this)` and that method *hides* `Unsafe` type...

both `SingleThread` and `AccordCommandStore` schedule a `Runnable` in their `Executor` which will call `CommandStore.register(this)`

Reply:
I mean that the `SingleThread` (and extensions) do not need to register, or use the `Unsafe` implementation at all, since their implementation of `inStore()` is:

```
        @Override
        public boolean inStore()
        {
            return thread == Thread.currentThread();
        }
```

However, the other `Synchronised` and `DelayedCommandStore` do need some help, and they rely on `Unsafe` to implement `inStore`... except that their `register` call on construction won't work in tandem with these calls.

Reply:
> I mean that the SingleThread (and extensions) do not need to register,

If we do not register, how can we fetch the current store? `accord.local.CommandStore#current` is used in `Node` to know which `CommandStore` is running so it can schedule the callback there.

> except that their register call on construction won't work in tandem with these calls.

Correct, which is why their `register` method is a `noop`.  Both need the `Unsafe` methods as they "weave" between stores using the same thread (main).

Reply:
Ah, I missed that this implements `current`. 

Why don't we make `CommandStore` an abstract class? It was before, and it will be again (Bootstrap does this to bring shared functionality into it). No need to have an `Unsafe`, just control it from there.

Also, we can safely register from within the constructor - we already "unsafe" publish the `CommandStore` into the thread to setup the `thread` field, this is no different, and it would be cleaner.

Reply:
> we already "unsafe" publish the CommandStore into the thread to setup the thread field

Are you talking about `SingleThread`'s 

```
executor.execute(() -> thread = Thread.currentThread());
```

This does not expose `this` for other callers to see, and uses a param rather than a field, so we don't "unsafe publish"

> this is no different, and it would be cleaner.

I disagree, this is different as we are exposing the field for others to consume.  My understanding of JMM is that this publish isn't guarantied to be visible in the new thread even after the constructor completes, which could lead to a task seeing a partial `CommandStore`.

How about this compromise... we take a `CommandStore.Factory` and not actually directly call `new`... so the `Thread` based `CommandStores` could do `new` + `register` in that factory so this doesn't leak into `CommandStores`?

Reply:
Pushed a change (see 12c558ea6cabdbb4beb42883545e1d7177006ca8) so `SingleThread` and `AccordCommandStore` both use a new `create` method that hides `new` and will register with the `ThreadLocal`.  This let me hide this detail from `CommandStores`

Reply:
moved `CommandStore` to an `abstract class` and removed `Unsafe`; the `runWith` methods are now `protected` methods 

see 69cc0b0aed695a50329e6fa90346bb9c74b5262c